Fixes #13274: Allow cygwin-paths on Windows

kvitorr opened the pull request:
Closes https://github.com/JabRef/jabref/issues/13274

### **Allow cygwin-paths on Windows**
Create the method **convertCygwinPathToWindows** in the FileUtil class. It transforms a file path from Cygwin-style to Windows-style only if the operating system is Windows.

Based on research, the two common Cygwin-style formats are:
- /cygdrive/{drive}/some/path
- /{drive}/some/path

The third was asked to be added
-  /mnt/{drive}/Users/ (asked in the code review)

The method follows this logic:
```java
/// Converts a Cygwin-style file path to a Windows-style path if the operating system is Windows.
    ///
    /// Supported formats:
    /// - /cygdrive/c/Users/... → C:\Users\...
    /// - /mnt/c/Users/...      → C:\Users\...
    /// - /c/Users/...          → C:\Users\...
    ///
    /// @param filePath the input file path
    /// @return the converted path if running on Windows and path is in Cygwin format; otherwise, returns the original path
    public static Path convertCygwinPathToWindows(String filePath) {
        if (filePath == null) {
            return null;
        }

        if (!OS.WINDOWS) {
            return Path.of(filePath);
        }

        if (filePath.startsWith(MNT_PREFIX) && filePath.length() > 5) {
            return buildWindowsPathWithDriveLetterIndex(filePath, 5);
        }

        if (filePath.startsWith(CYGDRIVE_PREFIX) && filePath.length() > 10) {
            return buildWindowsPathWithDriveLetterIndex(filePath, 10);
        }

        if (ROOT_DRIVE_PATTERN.matcher(filePath).matches()) {
            return buildWindowsPathWithDriveLetterIndex(filePath, 1);
        }

        return Path.of(filePath);
    }

    /// Builds a Windows-style path from a Cygwin-style path using a known prefix index.
    /// @param path the input file path
    /// @param letterIndex the index driver letter, zero-based indexing
    /// @return a windows-style path
    private static Path buildWindowsPathWithDriveLetterIndex(String path, int letterIndex) {
        String driveLetter = path.substring(letterIndex, letterIndex + 1).toUpperCase();
        String windowsPath = path.substring(letterIndex + 1).replace("/", "\\\\");
        return Path.of(driveLetter + ":" + windowsPath);
    }
```

Additionally, unit tests for this method were added to **FileUtilTest class**

OBS: I need help to run the second test when Windows is disabled... I am getting an error because the test is not recognized...

```java
    @DisabledOnOs(value = org.junit.jupiter.api.condition.OS.WINDOWS, disabledReason = "Test in others operational systems")
    @ParameterizedTest
    @ValueSource(strings = {"/home/username/Downloads/test.bib"})
    void convertCygwinPathToWindowsShouldReturnOriginalFilePathWhenRunningOnWindows(String filePath) {
        assertEquals(Path.of(filePath), FileUtil.convertCygwinPathToWindows(filePath));
    }
```

It was created a Converter, called CygwinPathConverter, as requested on the code review... and it was added to the Option Anottation in the classes: CheckConsistency.java, CheckIntegrity.java, Convert.java, Pseudonymize.java, Search.java

**Example of use:**
```java
@Option(names = {"--input"}, converter = CygWinPathConverter.class, description = "Input BibTeX file", required = true)
    private Path inputFile;
```


### Mandatory checks

<!--
Go through the checklist below. It is mandatory, even for a draft pull request.

Keep ALL the items. Replace the dots inside [.] and mark them as follows: 
[x] done 
[ ] not done 
[/] not applicable
-->

- [x] I own the copyright of the code submitted and I license it under the [MIT license](https://github.com/JabRef/jabref/blob/main/LICENSE)
- [x] Change in `CHANGELOG.md` described in a way that is understandable for the average user (if change is visible to the user)
- [x] Tests created for changes (if applicable)
- [ ] Manually tested changed features in running JabRef (always required)
- [/] Screenshots added in PR description (if change is visible to the user)
- [x] [Checked developer's documentation](https://devdocs.jabref.org/): Is the information available and up to date? If not, I outlined it in this pull request.
- [ ] [Checked documentation](https://docs.jabref.org/): Is the information available and up to date? If not, I created an issue at <https://github.com/JabRef/user-documentation/issues> or, even better, I submitted a pull request to the documentation repository.

jabref-machine commented:
Your pull request needs to link an issue correctly.

To ease organizational workflows, please link this pull-request to the issue with syntax as described in <https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue>:

> <h2 id="linking-a-pull-request-to-an-issue-using-a-keyword">Linking a pull request to an issue using a keyword</h2>
> <p>You can link a pull request to an issue by using a supported keyword
> in the pull request's description or in a commit message.

### Examples

- ✅ `Fixes #xyz` links pull-request to issue. Merging the PR will close the issue.
- ✅ `Fixes https://github.com/JabRef/jabref/issues/xyz` links pull-request to issue. Merging the PR will close the issue.
- ✅ `Fixes https://github.com/Koppor/jabref/issues/xyz` links pull-request to issue. Merging the PR will close the issue.
- ❌ `Fixes [#xyz](https://github.com/JabRef/jabref/issues/xyz)` links pull-request to issue. Merging the PR will **NOT** close the issue.


<!-- ghprcomment
On PR opened/updated
Determine issue number
-->

jabref-machine commented:
You have removed the "Mandatory Checks" section from your pull request description. Please adhere to our [pull request template](https://github.com/JabRef/jabref/blob/main/.github/PULL_REQUEST_TEMPLATE.md?plain=1#L10).


<!-- ghprcomment
Tests
Mandatory Checks present
-->

jabref-machine commented:
Note that your PR will not be reviewed/accepted until you have gone through the mandatory checks in the description and marked each of them them exactly in the format of `[x]` (done), `[ ]` (not done yet) or `[/]` (not applicable).


<!-- ghprcomment
Tests
PR checklist OK
-->

koppor commented on jablib/src/main/java/org/jabref/logic/util/io/FileUtil.java:
```diff
@@ -589,4 +590,34 @@ public static String shortenFileName(String fileName, Integer maxLength) {
         return Arrays.binarySearch(ILLEGAL_CHARS, c) < 0;
     }
+
+    /**
```
Please convert to Markdown (///) to have it properly rendering

koppor commented on jablib/src/main/java/org/jabref/logic/util/io/FileUtil.java:
```diff
@@ -589,4 +590,34 @@ public static String shortenFileName(String fileName, Integer maxLength) {
+     * Converts a Cygwin-style file path to a Windows-style path, if the operating system is Windows.
+     *
+     * Supported formats:
+     * - /cygdrive/c/Users/...   → C:\Users\...
```
Can you add /mnt/c, too?

I totally forgot while writing

koppor commented:
Please add a CHANGELOG.md entry 😅

kvitorr commented:
@koppor can you help me? I do not where to create the test for the ArgumentProcessor class. Also, do you want me to change the name of the test cases?

kvitorr commented:
> > @koppor, can you help me? I’m not sure where to create the test for the ArgumentProcessor class. Also, would you like me to rename the test cases?
> 
> No need to create tests there. Just implement the transformation of the --input (and --output) filename(s).

already done that in the last commit

koppor commented:
> > > > > @koppor, can you help me? I’m not sure where to create the test for the ArgumentProcessor class. Also, would you like me to rename the test cases?
> > > > 
> > > > 
> > > > No need to create tests there. Just implement the transformation of the --input (and --output) filename(s).
> > > 
> > > 
> > > already done that in the last commit
> > 
> > See my comment. Is it possible to move the logic towards the --import argument?
> > 
> > Reason: should also work at pseudonymization and check consistency, not only import. (If I understand the current diff right currently only at the import command it's working. Did you try out for pseudonymize?)
> 
> sorry... I didn't see your comment. I'll try to understand what are you saying and then I'll comeback with some commit or comment 

Maybe it got lost while I was on the road.

There is a place in the code where --input is mapped to a Java object variable by the picocli annotation @option (or similar). Maybe, there is a transformion possible.

jabref-machine commented:
Your code currently does not meet JabRef's code guidelines. We use [Gradle Modernizer Plugin](https://github.com/andygoossens/gradle-modernizer-plugin#gradle-modernizer-plugin) to ensure "modern" Java coding practices. You can see which checks are failing by locating the box "Some checks were not successful" on the pull request page. To see the test output, locate "Tests / Modernizer (pull_request)" and click on it.

Please fix the detected errors, commit, and push.


<!-- ghprcomment
Tests
Modernizer
-->

kvitorr commented on jabkit/src/main/java/module-info.java:
```diff
@@ -20,9 +21,8 @@
-
     requires java.xml;
-  
+
```
remove line

jabref-machine commented:
Your code currently does not meet JabRef's code guidelines. We use [Gradle Modernizer Plugin](https://github.com/andygoossens/gradle-modernizer-plugin#gradle-modernizer-plugin) to ensure "modern" Java coding practices. You can see which checks are failing by locating the box "Some checks were not successful" on the pull request page. To see the test output, locate "Tests / Modernizer (pull_request)" and click on it.

Please fix the detected errors, commit, and push.


<!-- ghprcomment
Tests
Modernizer
-->

koppor commented on jabkit/src/main/java/org/jabref/cli/Convert.java:
```diff
@@ -36,13 +38,13 @@ public class Convert implements Runnable {
     private String inputFormat;
 
-    @Option(names = {"--output"}, description = "Output file")
+    @Option(names = {"--output"}, converter = PathConverter.class, description = "Output file")
```
Strange that one os PathConverter and the other one is StringPathConverter.

OK, this seems to be generated with AI. Please, instruct the AI to avoid code duplication.

`PathConverter` is the nicer name, remove the other things.

While scrolling through, you should have seen this for yourself @kvitorr!

kvitorr replied:
@koppor yes, I am using AI to understand the code... I have never worked with picocli... The converters were made taking into account the type of data that was returned by --input and --output options... So, if there was code duplication it was because there was no consensus on the return type of these variables. I didn't feel comfortable changing the type without knowing the consequences, so I uploaded the changes and waited for the code review.

kvitorr commented:
@calixtus thank you for your time, I will do the changes and then wait for the feedback. Please be patient xD, I am doing my best.

kvitorr commented:
@koppor @calixtus 

I made new changes to the project...
* I changed the type of the --input and --output attributes from String and File to Path as requested.

* I removed the code duplication from the converters and changed the name of PathConverter to CygwinPathConverter so that no one gets the impression that it is an unnecessary conversion class, as pointed out by calixtus.

* I changed the ConvertCygwinPathToWindows method of the FileUtil class to return a Path instead of a String.

* I created three constants in the FileUtil class, two representing the prefixes "/mnt/" and "/cygdrive/"... the third constant represents the pattern /{driveLetter}/... which are being used within the conditionals.

* I created a private method within the FileUtil class that is being used within ConvertCygwinPathToWindows... this code snippet was being repeated within the conditionals... and since code duplication was something that was mentioned in the review, I thought it would be better to create this new method.

Other than that, I didn't manually test the changes but I wanted to upload the modification right away so you could give me feedback on whether I understood the requested changes correctly.

I also updated the pull request description.

---

Regarding the use of AI, I understand your point of view, but I didn't find your comments interesting.

I use it to understand the code, understand dependencies, understand business rules, make comparisons between technologies, understand differences between data types, and try to bring more context to some of the comments you make.

The pull request description was not made using AI, and I understood the changes that were made and also tested it locally.

Yes, there was code duplication, but I didn't understand whether the data type received by the CLI could be changed from String/File to Path. My thought at the time was "if the code is using different data types for input and output, there must be a reason." And that's why I created the three converters, String, File, and Path.

So I thought, I'll upload the change and wait for the review. If I need to change something, it will be requested.

Since this is my FIRST open source contribution, I expected it to be a cool experience. In fact, I'm learning a lot, having contact with organized code, with tests, with documentation and with Picocli, but these comments about the use of AI caught me by surprise.

I read Daniel Roe's text (https://roe.dev/blog/using-ai-in-open-source) and I feel like I didn't deviate from anything he said.

But since this was pointed out to me, I also want to point out something to you, that you at least read the entire description of the pull request and answer any questions written there... this will help clarify ideas, allow a better understanding of the issue and maintain a good environment for both of us.

And, as it is said in the contribution guide, you can reject the PR if you think it doesn't contribute anything to this issue.

calixtus commented:
Sorry you got offended by our comment about ai. It's partially a precaution since we get a lot of contributions lately that are completely ai generated, look nice on first sight, but are rubbish on second.

We are not against use of ai, in fact we use it ourselves for repeating task, for simple refactors etc. The general rule is: never to use generated stuff without reading it, testing it and owning it. In short: Never let an ai think for you. Never let an ai speak for you.

There were some indicators, that some of your changes were generated. Maybe it is also our perspective, that makes look changes a newcomer made without much experience so partially similar to changes a LLM does. Eg. Your comments remove line and add line looked like commands one wanted to send to an ai.

Please don't be alienated by our precautions. We're always happy about newcomers and try to be welcoming. This is why we did comment on your changes and not just close the pr without comment because we would have thought that this would just be another ai pr.

calixtus commented:
Looks good to me so far...
is this still a draft or ready for review?

koppor commented on jabkit/src/main/java/org/jabref/cli/converter/CygWinPathConverter.java:
```diff
@@ -0,0 +1,16 @@
+package org.jabref.cli.converter;
```
For naming, I think, this should be with "s", because there will be more than one converters in the future.

`pacakge org.jabref.cli.converters`.

koppor commented on jabkit/src/main/java/module-info.java:
```diff
@@ -22,7 +23,6 @@
     requires org.tinylog.impl;
 
     requires java.xml;
-  
```
Please do not modify lines of code not related to the PR if not 90% sure

Here, the empty line does make sence to separate blocks. Please undo this change.

koppor commented on jablib/src/main/java/org/jabref/logic/util/io/FileUtil.java:
```diff
@@ -589,4 +593,47 @@ public static String shortenFileName(String fileName, Integer maxLength) {
+    ///
+    /// @param filePath the input file path
+    /// @return the converted path if running on Windows and path is in Cygwin format; otherwise, returns the original path
+    public static Path convertCygwinPathToWindows(String filePath) {
```
The method also works on linux - proposal for another name:

```suggestion
    public static Path convertStringToPathRespecingCygwin(String filePath) {
```

koppor commented on jablib/src/main/java/org/jabref/logic/util/io/FileUtil.java:
```diff
@@ -589,4 +593,47 @@ public static String shortenFileName(String fileName, Integer maxLength) {
         return Arrays.binarySearch(ILLEGAL_CHARS, c) < 0;
     }
+
+    /// Converts a Cygwin-style file path to a Windows-style path if the operating system is Windows.
```
Add something like "plainly converted if on non-Windows"

koppor reviewed (changes requested):
Small comments.

Please also have a look at the failing jbang tests - maybe, some interface changed and needs to be adapted?

https://github.com/JabRef/jabref/actions/runs/15646674264/job/44085291757?pr=13297

koppor commented:
@kvitorr Will you continue here?

koppor reviewed (approved):
I think, we can fix this nitpicks later.

jabref-machine commented:
Note that your PR will not be reviewed/accepted until you have gone through the mandatory checks in the description and marked each of them them exactly in the format of `[x]` (done), `[ ]` (not done yet) or `[/]` (not applicable).


<!-- ghprcomment
PR Tests
PR checklist OK
-->

koppor reviewed (approved):
I think, its good enough

Future work might be to make it available for JabLS and JabSrv. CC @palukku
